app/src/main/java/com/asfoundation/wallet/ui/onboarding/OnboardingPresenter.kt

```diff
@@ -1,74 +1,138 @@
 package com.asfoundation.wallet.ui.onboarding
 
 import android.net.Uri
+import com.asfoundation.wallet.entity.Wallet
+import com.asfoundation.wallet.interact.SmsValidationInteract
+import com.asfoundation.wallet.wallet_validation.WalletValidationStatus
+import io.reactivex.Completable
 import io.reactivex.Observable
 import io.reactivex.Scheduler
 import io.reactivex.disposables.CompositeDisposable
 import io.reactivex.functions.BiFunction
+import io.reactivex.subjects.PublishSubject
 import java.util.concurrent.TimeUnit
 
 class OnboardingPresenter(private val disposables: CompositeDisposable,
                           private val view: OnboardingView,
                           private val onboardingInteract: OnboardingInteract,
-                          private val viewScheduler: Scheduler) {
+                          private val viewScheduler: Scheduler,
+                          private val smsValidationInteract: SmsValidationInteract,
+                          private val networkScheduler: Scheduler) {
+
+  private var walletCreated: PublishSubject<Boolean> = PublishSubject.create()
 
   fun present() {
-    handleOnBoardingFinish()
+    handleSkipedOnboarding()
     handleLinkClick()
+    handleCreateWallet()
+    handleRedeemButtonClicks()
+    handleNextButtonClicks()
+    handleLaterClicks()
+    handleRetryClicks()
   }
 
   fun stop() {
     disposables.clear()
   }
 
-  private fun handleSkip(): Observable<Any> {
-    return Observable.fromCallable { onboardingInteract.hasClickedSkipOnboarding() }
-        .flatMap {
-          if (it) {
-            view.showLoading()
-            return@flatMap Observable.just(true)
-          }
-          return@flatMap handleSkipButtonClick()
-        }
+  private fun isWalletCreated(): Observable<Boolean> {
+    return walletCreated.filter { created -> created }
   }
 
-  private fun handleSkipButtonClick(): Observable<Any> {
-    return view.getSkipButtonClick()
-        .doOnNext {
-          onboardingInteract.clickSkipOnboarding()
-          view.showLoading()
-        }
+  private fun handleRetryClicks() {
+    disposables.add(
+        Observable.combineLatest(isWalletCreated(), view.getRetryButtonClicks(),
+            BiFunction { _: Boolean, _: Any -> }
+        )
+            .flatMapSingle { onboardingInteract.getWalletAddress() }
+            .flatMapSingle {
+              smsValidationInteract.isValid(Wallet(it))
+                  .subscribeOn(networkScheduler)
+            }
+            .delay(1, TimeUnit.SECONDS)
```

Comment:
Question: Since we have a delay on the retry click on the view, doesn’t this add up to 2 seconds?